File: object_graph_streamer_test.go
package ObjectGraphStreamer

import (
	"crypto/sha256"
	"testing"
	"time"

	"github.com/mabels/object-graph-streamer/mocks"

	"github.com/btcsuite/btcutil/base58"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/suite"
)

type ObjectGraphStreamSuite struct {
	suite.Suite
	mockedSvalFn *SvalFnMock
}

func (s *ObjectGraphStreamSuite) SetupTest() {
	s.mockedSvalFn = &SvalFnMock{}
}

// ####################
// ## ObjectGraphStreamer tests ##
// ####################

func (s *ObjectGraphStreamSuite) TestSortWithOutWithString() {
	strValue := "string"
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		return assert.Equal(s.T(), strValue, (sVal.Val.(JsonValType)).Val)
	}))
	ObjectGraphStreamer(strValue, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithDate() {
	t := time.Now()
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		return assert.Equal(s.T(), t, (sVal.Val.(JsonValType)).Val)
	}))
	ObjectGraphStreamer(t, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithNumber() {
	n := 78
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		return assert.Equal(s.T(), n, (sVal.Val.(JsonValType)).Val)
	}))
	ObjectGraphStreamer(n, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithBoolean() {
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		return assert.Equal(s.T(), true, (sVal.Val.(JsonValType)).Val)
	}))
	ObjectGraphStreamer(true, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithArrayOfEmpty() {
	var emptySlice []int
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_START, sVal.OutState.String())
		} else if funcCallIdx == 2 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_END, sVal.OutState.String())
		}
		funcCallIdx++
		return result
	}))
	ObjectGraphStreamer(emptySlice, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithArrayOf_1_2() {
	ar := []int{1, 2}
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_START, sVal.OutState.String())
		} else if funcCallIdx == 2 {
			result = assert.Equal(s.T(), ar[0], (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 3 {
			result = assert.Equal(s.T(), ar[1], (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 4 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_END, sVal.OutState.String())
		}
		funcCallIdx++
		return result
	}))
	ObjectGraphStreamer(ar, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithArrayOf_1_2_3_4() {
	ar := [][]int{{1, 2}, {3, 4}}
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 || funcCallIdx == 2 || funcCallIdx == 6 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_START, sVal.OutState.String())
		} else if funcCallIdx == 3 {
			result = assert.Equal(s.T(), ar[0][0], (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 4 {
			result = assert.Equal(s.T(), ar[0][1], (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 5 || funcCallIdx == 9 || funcCallIdx == 10 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), ARRAY_END, sVal.OutState.String())
		} else if funcCallIdx == 7 {
			result = assert.Equal(s.T(), ar[1][0], (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 8 {
			result = assert.Equal(s.T(), ar[1][1], (sVal.Val.(JsonValType)).Val)
		}
		funcCallIdx++
		return result
	}))
	ObjectGraphStreamer(ar, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithObjOfEmptyObj() {
	var obj struct{}
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_START, sVal.OutState.String())
		} else if funcCallIdx == 2 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_END, sVal.OutState.String())
		}
		funcCallIdx++
		return result
	}))
	ObjectGraphStreamer(obj, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithObjOfObj_Y_1_X_2() {
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_START, sVal.OutState.String())
		} else if funcCallIdx == 2 {
			result = assert.Equal(s.T(), "x", sVal.Attribute)
		} else if funcCallIdx == 3 {
			result = assert.Equal(s.T(), 2, (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 4 {
			result = assert.Equal(s.T(), "y", sVal.Attribute)
		} else if funcCallIdx == 5 {
			result = assert.Equal(s.T(), 1, (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 6 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_END, sVal.OutState.String())
		}
		funcCallIdx++
		return result
	}))
	ObjectGraphStreamer(struct {
		Y int `json:"y"`
		X int `json:"x"`
	}{Y: 1, X: 2}, s.mockedSvalFn.Execute)
}

func (s *ObjectGraphStreamSuite) TestSortWithOutWithObjOfObj_Y_B_1_A_2() {
	funcCallIdx := 1
	s.mockedSvalFn.On("Execute", mock.MatchedBy(func(sVal SVal) bool {
		result := false
		if funcCallIdx == 1 || funcCallIdx == 3 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_START, sVal.OutState.String())
		} else if funcCallIdx == 2 {
			result = assert.Equal(s.T(), "y", sVal.Attribute)
		} else if funcCallIdx == 4 {
			result = assert.Equal(s.T(), "a", sVal.Attribute)
		} else if funcCallIdx == 5 {
			result = assert.Equal(s.T(), 2, (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 6 {
			result = assert.Equal(s.T(), "b", sVal.Attribute)
		} else if funcCallIdx == 7 {
			result = assert.Equal(s.T(), 1, (sVal.Val.(JsonValType)).Val)
		} else if funcCallIdx == 8 || funcCallIdx == 9 {
			result = assert.Equal(s.T(), nil, sVal.Val)
			result = result && assert.Equal(s.T(), OBJECT_END, sVal.OutState.String())
		}
		funcCallIdx++
		return result
	}))

	type Obj struct {
		B int `json:"b"`
		A int `json:"a"`
	}
	ObjectGraphStreamer(struct {
		Y Obj `json:"y"`
	}{Y: Obj{
		B: 1,
		A: 2,
	}}, s.mockedSvalFn.Execute)
}

// #########################
// ## JSONCollector tests ##
// #########################
func (s *ObjectGraphStreamSuite) TestJSONCollectorEmptyObj() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	var obj struct{}
	ObjectGraphStreamer(obj, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "{}", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorEmptyArray() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	var emptySlice []int
	ObjectGraphStreamer(emptySlice, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollector_X_Y_1_Z_x_Y_Z() {
	type Obj struct {
		Y int    `json:"y"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}

	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	ObjectGraphStreamer(struct {
		X Obj      `json:"x"`
		Y struct{} `json:"y"`
		Z []int    `json:"z"`
	}{
		X: Obj{
			Y: 1,
			Z: "x",
		},
		Y: emptypObj,
		Z: emptySlice,
	}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "{\"x\":{\"y\":1,\"z\":\"x\"},\"y\":{},\"z\":[]}", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorArray_xx() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	ObjectGraphStreamer([]string{"xx"}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[\"xx\"]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorArray_1_2() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	ObjectGraphStreamer([]interface{}{1, "2"}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[1,\"2\"]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollector_1_2_A() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	ObjectGraphStreamer([]interface{}{1, []string{"2", "A"}, "E"}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[1,[\"2\",\"A\"],\"E\"]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorIndent2EmptyObj() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, NewJsonProps(2, ""))
	var obj struct{}
	ObjectGraphStreamer(obj, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "{}", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorIndent2ArrayEmpty() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, NewJsonProps(2, ""))
	var emptySlice []int
	ObjectGraphStreamer(emptySlice, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollectorIndent2_X_Y_1_Z_x() {
	type Obj struct {
		Y int    `json:"y"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}

	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, NewJsonProps(2, ""))
	ObjectGraphStreamer(struct {
		X Obj      `json:"x"`
		Y struct{} `json:"y"`
		Z []int    `json:"z"`
	}{
		X: Obj{
			Y: 1,
			Z: "x",
		},
		Y: emptypObj,
		Z: emptySlice,
	}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "{\n  \"x\": {\n    \"y\": 1,\n    \"z\": \"x\"\n  },\n  \"y\": {},\n  \"z\": []\n}", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollector_Indent2_xx() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, NewJsonProps(2, ""))
	ObjectGraphStreamer([]string{"xx"}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[\n  \"xx\"\n]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollector_Indent2_array_1_2() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, NewJsonProps(2, ""))
	ObjectGraphStreamer([]interface{}{1, "2"}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[\n  1,\n  \"2\"\n]", out)
}

func (s *ObjectGraphStreamSuite) TestJSONCollector_1_date444() {
	out := ""
	col := NewJsonCollector(func(str string) {
		out += str
	}, nil)
	ObjectGraphStreamer([]interface{}{1, time.UnixMilli(444).UTC()}, func(prob SVal) {
		col.Append(prob)
	})
	assert.Equal(s.T(), "[1,\"1970-01-01T00:00:00.444Z\"]", out)
}

// #########################
// ## HashCollector tests ##
// #########################
func (s *ObjectGraphStreamSuite) TestHashCollector_date() {
	h := NewHashCollector()
	ObjectGraphStreamer(time.UnixMilli(444).UTC(), func(prob SVal) {
		h.Append(prob)
	})
	assert.Equal(s.T(), "DzYqv3YaniBJWwqrNBn4534oTe4nL14TqcfVCguf9Yyv", h.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_X_1_Y2() {
	h := NewHashCollector()
	ObjectGraphStreamer(struct {
		X int
		Y int
	}{
		X: 1,
		Y: 2,
	}, func(prob SVal) {
		h.Append(prob)
	})
	assert.Equal(s.T(), "DkPs9C3fYabdDLFxqMh4ZoTNr3xD1xYGFvYnJioF7V6H", h.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_1() {
	h := NewHashCollector()
	type Obj struct {
		Y int    `json:"y"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}
	ObjectGraphStreamer(struct {
		X Obj       `json:"x"`
		Y struct{}  `json:"y"`
		Z []int     `json:"z"`
		D time.Time `json:"d"`
	}{
		X: Obj{
			Y: 1,
			Z: "x",
		},
		Y: emptypObj,
		Z: emptySlice,
		D: time.UnixMilli(444).UTC(),
	}, func(prob SVal) {
		h.Append(prob)
	})
	assert.Equal(s.T(), "5PvJAWGkaKAHax6tsaKGfPYm6JfXxZs15wRTDpSKaZ2G", h.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_2() {
	h := NewHashCollector()
	type Obj struct {
		Y int    `json:"y"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}
	ObjectGraphStreamer(struct {
		X    Obj       `json:"x"`
		Y    struct{}  `json:"y"`
		Z    []int     `json:"z"`
		Date time.Time `json:"date"`
	}{
		X: Obj{
			Y: 2,
			Z: "x",
		},
		Y:    emptypObj,
		Z:    emptySlice,
		Date: time.UnixMilli(444).UTC(),
	}, func(prob SVal) {
		h.Append(prob)
	})
	assert.Equal(s.T(), "ECVWfmcNaUGkgvPZe7CojrnRNULxNczKXU8PGns6UDvr", h.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_3() {
	h := NewHashCollector()
	type Obj struct {
		X int    `json:"x"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}
	ObjectGraphStreamer(struct {
		X    Obj       `json:"x"`
		Y    struct{}  `json:"y"`
		Z    []int     `json:"z"`
		Date time.Time `json:"date"`
	}{
		X: Obj{
			X: 1,
			Z: "x",
		},
		Y:    emptypObj,
		Z:    emptySlice,
		Date: time.UnixMilli(444).UTC(),
	}, func(prob SVal) {
		h.Append(prob)
	})
	assert.Equal(s.T(), "EoYNGMtap1k9iEAGeVtHmJwpMjQLKWJmR27SG6aC9fSg", h.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_4() {
	h1 := NewHashCollector()
	type Obj struct {
		X int    `json:"x"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}
	ObjectGraphStreamer(struct {
		X    Obj       `json:"x"`
		Y    struct{}  `json:"y"`
		Z    []int     `json:"z"`
		Date time.Time `json:"date"`
	}{
		X: Obj{
			X: 1,
			Z: "x",
		},
		Y:    emptypObj,
		Z:    emptySlice,
		Date: time.UnixMilli(444).UTC(),
	}, func(prob SVal) {
		h1.Append(prob)
	})

	h2 := NewHashCollector()
	ObjectGraphStreamer(struct {
		Date time.Time `json:"date"`
		X    Obj       `json:"x"`
		Y    struct{}  `json:"y"`
		Z    []int     `json:"z"`
	}{
		Date: time.UnixMilli(444).UTC(),
		X: Obj{
			X: 1,
			Z: "x",
		},
		Y: emptypObj,
		Z: emptySlice,
	}, func(prob SVal) {
		h2.Append(prob)
	})

	assert.Equal(s.T(), h1.Digest(), h2.Digest())
}

func (s *ObjectGraphStreamSuite) TestHashCollector_3_InternalUpdate() {
	hashCalculator := sha256.New()

	type Obj struct {
		R int    `json:"r"`
		Z string `json:"z"`
	}
	var emptySlice []int
	var emptypObj struct{}
	expectedArgs := []string{"date", "1970-01-01T00:00:00.444Z", "x", "r", "1", "z", "u", "y", "z"}
	mck := &mocks.Hash{}
	idx := 0
	mck.On("Write", mock.MatchedBy(func(p []byte) bool {
		hashCalculator.Write(p)
		idx++
		return expectedArgs[idx-1] == string(p)
	})).Return(1, nil)

	t := struct {
		X    Obj       `json:"x"`
		Y    struct{}  `json:"y"`
		Z    []int     `json:"z"`
		Date time.Time `json:"date"`
	}{
		X: Obj{
			R: 1,
			Z: "u",
		},
		Y:    emptypObj,
		Z:    emptySlice,
		Date: time.UnixMilli(444).UTC(),
	}
	collector := &HashCollector{mck}
	ObjectGraphStreamer(t, func(prob SVal) {
		collector.Append(prob)
	})

	var nilBytes []byte
	mck.On("Sum", nilBytes).Return([]byte{})
	collector.Digest()
	mck.AssertNumberOfCalls(s.T(), "Sum", 1)

	assert.Equal(s.T(), "CwEMjUHV6BpDS7AGBAYqjY6qMKE6xC8Z56H5T2ZuUuXe", base58.Encode(hashCalculator.Sum(nil)))
}

func (s *ObjectGraphStreamSuite) TestSimpleHash() {
	type Data struct {
		Name string `json:"name"`
		Date string `json:"date"`
	}

	type KindData struct {
		Kind string `json:"kind"`
		Data Data   `json:"data"`
	}

	hashC := NewHashCollector()
	ObjectGraphStreamer(KindData{
		Kind: "test",
		Data: Data{
			Name: "object",
			Date: "2021-05-20",
		},
	}, func(sval SVal) {
		hashC.Append(sval)
	})
	assert.Equal(s.T(), "5zWhdtvKuGob1FbW9vUGPQKobcLtYYr5wU8AxQRVraeB", hashC.Digest())
}

// ##########################
// ## SimpleEnvelope tests ##
// ##########################

func TestObjectGraphStreamerSuite(t *testing.T) {
	suite.Run(t, new(ObjectGraphStreamSuite))
}

type SvalFnMock struct {
	mock.Mock
}

// Execute provides a mock function with given fields: prob
func (_m *SvalFnMock) Execute(prob SVal) {
	_m.Called(prob)
}

File: src/object_graph_streamer_test.py
from datetime import datetime, timezone, tzinfo
import hashlib
import json
import unittest
import unittest.mock

from object_graph_streamer import HashCollector, JsonCollector, JsonProps, objectGraphStreamer, OutState


class Mockdatetime:
    def now(self):
        return datetime.fromtimestamp(1624140000)


mockdatetime = Mockdatetime()


def toSVals(calls: list[unittest.mock.call]):
    return list(map(lambda x: list(map(lambda x: x.to_dict(), x.args)), calls))


class ObjectGraphStreamerTest(unittest.TestCase):

    def test_simple_hash(self):
        hashC = HashCollector()
        objectGraphStreamer({
            'kind': "test",
            'data': {
                'name': "object",
                'date': "2021-05-20",
            },
        },
            lambda sval: hashC.append(sval))
        self.assertEqual(
            hashC.digest(), "5zWhdtvKuGob1FbW9vUGPQKobcLtYYr5wU8AxQRVraeB")

    def test_sort_with_out_with_string(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer("string", fn)
        self.assertEqual(toSVals(fn.mock_calls), [
                         [{'val': {'val': "string", }, 'outState': 'V', 'paths': []}]])

    def test_sort_with_out_with_date(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer(datetime.fromtimestamp(0.444, tz=timezone.utc), fn)
        self.assertEqual(toSVals(fn.mock_calls),
                         [[{'val': {'val': datetime.fromtimestamp(0.444, tz=timezone.utc)}, 'outState': 'V', 'paths': []}]])

    def test_sort_with_out_with_number(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer(4711, fn)
        self.assertEqual(toSVals(fn.mock_calls), [
                         [{'val': {'val': 4711}, 'outState': 'V', 'paths': []}]])

    def test_sort_with_out_with_boolean(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer(False, fn)
        self.assertEqual(toSVals(fn.mock_calls), [
                         [{'val': {'val': False}, 'outState': 'V', 'paths': []}]])

    def test_sort_with_out_with_array_of_empty(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer([], fn)
        self.assertEqual(toSVals(fn.mock_calls),
                         [[{'outState': "[", 'paths': ['[']}], [{'outState': "]", 'paths': [']']}]])

    def test_sort_with_out_with_array_of_1_2(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer([1, 2], fn)
        self.assertEqual(toSVals(fn.mock_calls), [
            [{'outState': "[", 'paths': ['[']}],
            [{'val': {'val': 1}, 'outState': 'V', 'paths': ['[', '0']}],
            [{'val': {'val': 2}, 'outState': 'V', 'paths': ['[', '1']}],
            [{'outState': "]", 'paths': [']']}],
        ])

    def test_sort_with_out_with_array_of_1_2_3_4(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer(
            [
                [1, 2],
                [3, 4],
            ],
            fn
        )
        self.assertEqual(toSVals(fn.mock_calls), [
            [{'outState': "[", 'paths': ['[']}],
            [{'outState': "[", 'paths': ['[', '0', '[']}],
            [{'val': {'val': 1}, 'outState': 'V','paths': ['[', '0', '[', '0']}],
            [{'val': {'val': 2}, 'outState': 'V','paths': ['[', '0', '[', '1']}],
            [{'outState': "]", 'paths': ['[', '0', ']']}],
            [{'outState': "[", 'paths': ['[', '1', '[']}],
            [{'val': {'val': 3,}, 'outState': 'V', 'paths': ['[', '1', '[', '0']}],
            [{'val': {'val': 4}, 'outState': 'V', 'paths': ['[', '1', '[', '1']}],
            [{'outState': "]", 'paths': ['[', '1', ']']}],
            [{'outState': "]", 'paths': [']']}],
        ])

    def test_sort_with_out_with_obj_of_empty_obj(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer({}, fn)
        self.assertEqual(toSVals(fn.mock_calls),
                         [[{'outState': "{",  'paths': ['{']}], [{'outState': "}", 'paths': ['}']}]])

    def test_sort_with_out_with_obj_of_obj_y_1_x_2(self):
        fn = unittest.mock.Mock()
        objectGraphStreamer({'y': 1, 'x': 2}, fn)
        self.assertEqual(toSVals(fn.mock_calls), [
            [{'outState': "{", 'paths': ['{']}],
            [{'attribute': "x", 'outState': 'A', 'paths': ['{', 'x']}],
            [{'val': {'val': 2}, 'outState': 'V', 'paths': ['{', 'x']}],
            [{'attribute': "y", 'outState': 'A', 'paths': ['{', 'y']}],
            [{'val': {'val': 1}, 'outState': 'V', 'paths': ['{', 'y']}],
            [{'outState': "}", 'paths': ['}']}],
        ])

    def test_sort_with_out_with_obj_of_obj_y_b_1_a_2(self):
        self.maxDiff = None
        fn = unittest.mock.Mock()
        objectGraphStreamer({'y': {'b': 1, 'a': 2}}, fn)
        self.assertEqual(toSVals(fn.mock_calls), [
            [{'outState': "{", 'paths': ['{']}],
            [{'attribute': "y", 'outState': 'A', 'paths': ['{', 'y']}],
            [{'outState': "{", 'paths': ['{', 'y', '{']}],
            [{'attribute': "a", 'outState': 'A', 'paths': ['{', 'y', '{', 'a']}],
            [{'val': {'val': 2}, 'outState': 'V', 'paths': ['{', 'y', '{', 'a']}],
            [{'attribute': "b", 'outState': 'A', 'paths': ['{', 'y', '{', 'b']}],
            [{'val': {'val': 1}, 'outState': 'V', 'paths': ['{', 'y', '{', 'b']}],
            [{'outState': "}", 'paths': ['{', 'y', '}']}],
            [{'outState': "}", 'paths': ['}']}],
        ])

    def test_JSONCollector_empty_obj(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer({}, lambda o: json.append(o))
        self.assertEqual("".join(out), "{}")

    def test_JSONCollector_empty_array(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer([], lambda o: json.append(o))
        self.assertEqual("".join(out), "[]")

    def test_JSONCollector_x_y_1_z_x_y_z(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer({'x': {'y': 1, 'z': "x"}, 'y': {}, 'z': []},
                            lambda o: json.append(o))
        self.assertEqual("".join(out), '{"x":{"y":1,"z":"x"},"y":{},"z":[]}')

    def test_JSONCollector_array_xx(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer(["xx"], lambda o: json.append(o))
        self.assertEqual("".join(out), '["xx"]')

    def test_JSONCollector_array_1_2(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer([1, "2"], lambda o: json.append(o))
        self.assertEqual("".join(out), '[1,"2"]')

    def test_JSONCollector_1_2_A(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        objectGraphStreamer([1, ["2", "A"], "E"], lambda o: json.append(o))
        self.assertEqual("".join(out), '[1,["2","A"],"E"]')

    def test_JSONCollector_indent_2_empty_obj(self):
        out = []
        json = JsonCollector(lambda o: out.append(o),
                             JsonProps(**{'indent': 2}))
        objectGraphStreamer({}, lambda o: json.append(o))
        self.assertEqual("".join(out), "{}")

    def test_JSONCollector_indent_2_array_empty(self):
        out = []
        json = JsonCollector(lambda o: out.append(o),
                             JsonProps(**{'indent': 2}))
        objectGraphStreamer([], lambda o: json.append(o))
        self.assertEqual("".join(out), "[]")

    def test_JSONCollector_indent_2_x_y_1_z_x(self):
        out = []
        json = JsonCollector(lambda o: out.append(o),
                             JsonProps(**{'indent': 2}))
        objectGraphStreamer({'x': {'y': 1, 'z': "x"}, 'y': {}, 'z': []},
                            lambda o: json.append(o))
        self.assertEqual(
            "".join(out), '{\n  "x": {\n    "y": 1,\n    "z": "x"\n  },\n  "y": {},\n  "z": []\n}')

    def test_JSONCollector_indent_2_xx(self):
        out = []
        json = JsonCollector(lambda o: out.append(o),
                             JsonProps(**{'indent': 2}))
        objectGraphStreamer(["xx"], lambda o: json.append(o))
        self.assertEqual("".join(out), '[\n  "xx"\n]')

    def test_JSONCollector_indent_2_array_1_2(self):
        out = []
        json = JsonCollector(lambda o: out.append(o),
                             JsonProps(**{'indent': 2}))
        objectGraphStreamer([1, "2"], lambda o: json.append(o))
        self.assertEqual("".join(out), '[\n  1,\n  "2"\n]')

    def test_JSONCollector_1_date444(self):
        out = []
        json = JsonCollector(lambda o: out.append(o))
        obj = [1, datetime.fromtimestamp(0.444, tz=timezone.utc)]
        objectGraphStreamer(obj, lambda o: json.append(o))
        self.assertEqual("[1,\"1970-01-01T00:00:00.444Z\"]", "".join(out))

    def test_HashCollector_1(self):
        hash = HashCollector()
        objectGraphStreamer({'x': {'y': 1, 'z': "x"}, 'y': {}, 'z': [],
                             'd': datetime.fromtimestamp(0.444, tz=timezone.utc)}, lambda o: hash.append(o))
        self.assertEqual(
            hash.digest(), "5PvJAWGkaKAHax6tsaKGfPYm6JfXxZs15wRTDpSKaZ2G")

    def test_HashCollector_2(self):
        hash = HashCollector()
        objectGraphStreamer({'x': {'y': 2, 'z': "x"}, 'y': {}, 'z': [],
                             'date': datetime.fromtimestamp(0.444, tz=timezone.utc)}, lambda o: hash.append(o))
        self.assertEqual(
            hash.digest(), "ECVWfmcNaUGkgvPZe7CojrnRNULxNczKXU8PGns6UDvr")

    def test_HashCollector_3(self):
        hash = HashCollector()
        objectGraphStreamer({'x': {'x': 1, 'z': "x"}, 'y': {}, 'z': [],
                             'date': datetime.fromtimestamp(0.444, tz=timezone.utc)}, lambda o: hash.append(o))
        self.assertEqual(
            hash.digest(), "EoYNGMtap1k9iEAGeVtHmJwpMjQLKWJmR27SG6aC9fSg")

    def test_HashCollector_4(self):
        hash1 = HashCollector()
        objectGraphStreamer({'x': {'x': 1, 'z': "x"}, 'y': {}, 'z': [],
                             'date': datetime.fromtimestamp(0.444, tz=timezone.utc)}, lambda o: hash1.append(o))
        hash2 = HashCollector()
        objectGraphStreamer({'date': datetime.fromtimestamp(0.444, tz=timezone.utc), 'x': {
            'x': 1, 'z': "x"}, 'y': {}, 'z': []}, lambda o: hash2.append(o))
        self.assertEqual(hash1.digest(), hash2.digest())

    def test_HashCollector_3_internal_update(self):
        class MockHash:
            hash: any
            mockUpdate: any

            def __init__(self) -> None:
                self.mockUpdate = unittest.mock.Mock()
                self.hash = hashlib.new('sha256')

            def update(self, a: any):
                self.mockUpdate(a)
                self.hash.update(a)

            def digest(self):
                return self.hash.digest()

        hashCollector = HashCollector(MockHash())
        objectGraphStreamer({'x': {'r': 1, 'z': "u"}, 'y': {}, 'z': [], 'date': datetime.fromtimestamp(
            0.444, tz=timezone.utc)}, lambda o: hashCollector.append(o))
        result = list(map(lambda m: m.args[0].decode(
        ), hashCollector.hash.mockUpdate.mock_calls))
        self.assertEqual(result,
                         ["date", "1970-01-01T00:00:00.444Z", "x", "r", "1", "z", "u", "y", "z"])
        self.assertEqual(hashCollector.digest(),
                         "CwEMjUHV6BpDS7AGBAYqjY6qMKE6xC8Z56H5T2ZuUuXe")


if __name__ == '__main__':
    unittest.main()
